sos/plugins/juju_controller.py

```diff
@@ -0,0 +1,193 @@
+            "resources",
+            "sequence",
+            "settings",
+            "sshhostkeys",
+            "statuses",
+            "storageattachments",
+            "storageconstraints",
+            "storageinstances",
+            "storedResources",
+            "subnets",
+            "toolsmetadata",
+            "txns.prune",
+            "txns.stash",
+            "units",
+            "userLastLogin",
+            "usermodelname",
+            "users",
+            "volumes",
+        ]
+        # We don't collect txns.log, statuseshistory, metrics, txns due to size
+
+        for collection in collections:
+            filename = "mongoexport_collection_{}.json".format(collection)
+            # This will execute a mongoexport command, note it will add the
+            # password to the sos log
+            mongocmd = (
+                "mongoexport --host 127.0.0.1 --port 37017 --db juju "
+                + "--authenticationDatabase admin --ssl "
+                + "--sslAllowInvalidCertificates --username '{}' --password "
+                + "'{}' --collection {} --jsonArray"
+            )
+            self.add_cmd_output(
+                mongocmd.format(machineid, dbpass, collection),
+                suggest_filename=filename,
+            )
+
+    def get_machine_id(self):
+        # get the juju machine name machine-X
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                match = re.match(r"(machine-\d+)", dirname)
+                if match:
+                    return match.group(1)
+        return None
+
+    def get_jujudb_pass(self, machineid):
+        # Get agent config and figure out password
+        try:
+            fp = open("/var/lib/juju/agents/" + machineid + "/agent.conf", "r")
+            filecontents = fp.read()
+            fp.close()
+        except Exception:
+            return None
+        for line in iter(filecontents.splitlines()):
+            match = re.match(r"^statepassword: (\S+)", line)
+            if match:
+                return match.group(1)
+        return None
+
+    def add_query(self, machineid, dbpass, queryname, query):
+        # Run a command for a single query
+        filename = "mongo_{}.json".format(queryname)
+        # This will execute a mongo command, note it will add the password to
+        # the sos log
+        mongocmd = (
+            "mongo 127.0.0.1:37017/juju --authenticationDatabase admin"
+            + " --ssl --sslAllowInvalidCertificates --quiet --username '{}' "
+            + "--password '{}' --eval '{}'"
+        )
+        self.add_cmd_output(
+            mongocmd.format(machineid, dbpass, query),
+            suggest_filename=filename,
+        )
+
+    def setup(self):
+        self.add_copy_spec("/etc/default/mongodb")
+        self.add_copy_spec("/var/lib/juju/bootstrap-params")
+        self.add_copy_spec("/lib/systemd/system/juju-db/juju-db.service")
```

Comment:
Don't make repeated calls to `add_copy_spec()`; it accepts either a string, or a list of strings.